### PR TITLE
Elements: Avoid specificity bump for top-level element-only selectors

### LINKED ISSUE #63345: Rogues styles messing with my navigation styles
### Description

I have a rogue style overriding my navigation styles, namely underlines.

I do not have an underline on my navigation bar, but a very high-specificity style is being added (see attached screenshot):

`:root :where(a:where(:not(.wp-element-button)))`

My only solution is to add `!important` to my own styles. **Who decided that all `<a>` elements that are not buttons should be underlined?!!!**

Screenshot 2024-07-10 100749

### Step-by-step reproduction instructions

1. Install WordPress
2. Install Gutenberg

### Screenshots, screen recording, code snippet

_No response_

### Environment info

Latest WordPress, latest Gutenberg plugin, PHP 8.3+, Apache and Nginx, Windows 10 and Windows 11.

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

### PR DESCRIPTION
**Pull request by aaronrobertshaw**
Fixes: https://github.com/WordPress/gutenberg/issues/63345

## What?

With the levelling of CSS specificity in https://github.com/WordPress/gutenberg/pull/61638, it resulted in a bump in specificity for top level element-only selectors e.g. links, h1-6, etc. This PR avoids wrapping such selectors to maintain the previous specificity in <=6.5.

## Why?

If themes defined element styles that used only elements in the selector, the `0-1-0` from the `:root :where()` wrapped selectors would override them. This allows selectors such as `header nav ul li a` to continue to take precedence over our default core top level element styles.

## How?

Avoids wrapping the top-level element selectors in `:root :where()` unless those element selectors already used class level selectors. Pseudo selectors, e.g. `:hover` or `:focus`, are class-level selectors as well and this PR also makes sure they are capped at `0-1-0` for correct nesting of block style variations etc.

## Testing Instructions

After each step below, make sure that the same styles are applied on the frontend.

1. Add test nav structure/markup using custom HTML block
```html
<header>
  <nav>
     <ul>
       <li>
         <a>Test Link</a>
       </li>
     </ul>
  </nav>
</header>
```
2. Test custom global element styles including hover styles; Global Styles > Colors > Link 
3. Add custom css using element only selector `header nav ul li a`. You'll need to toggle the custom HTML block preview for it to pick up the custom CSS.
```css
header nav ul li a {
    color: red;
    text-decoration: none;
}
```
4. Test global block type element styles; Global Styles > Blocks > Group > Elements > Link
5. Test element style overrides on block instances
6. Define multiple block style variations that define both `styles.elements` and `styles.blocks.core/group.elements.link` styles. The following snippets can be used to create theme.json partial files within the theme's `/styles` directory which will then be registered as block style variations.
<details>
<summary>Example Block Style Variation Partials</summary>

```json
{
	"$schema": "https://schemas.wp.org/trunk/theme.json",
	"version": 3,
	"title": "Custom Links",
	"slug": "custom-links",
	"blockTypes": [ "core/group" ],
	"styles": {
		"elements": {
			"link": {
				"color": {
					"text": "crimson"
				},
				":hover": {
					"color": {
						"text": "darkmagenta"
					}
				}
			},
			"button": {
				"color": {
					"background": "purple",
					"text": "white"
				},
				":hover": {
					"color": {
						"background": "darkorchid"
					}
				}
			}
		},
		"blocks": {
			"core/group": {
				"elements": {
					"link": {
						"color": {
							"text": "dodgerblue"
						},
						":hover": {
							"color": {
								"text": "darkseagreen"
							}
						}
					},
					"heading": {
						"color": {
							"text": "teal"
						}
					}
				}
			}
		}
	}
}


```

```json
{
	"$schema": "https://schemas.wp.org/trunk/theme.json",
	"version": 3,
	"title": "Alt Links",
	"slug": "alt-links",
	"blockTypes": [ "core/group" ],
	"styles": {
		"elements": {
			"link": {
				"color": {
					"text": "green"
				},
				":hover": {
					"color": {
						"text": "lime"
					}
				}
			}
		},
		"blocks": {
			"core/group": {
				"elements": {
					"link": {
						"color": {
							"text": "coral"
						},
						":hover": {
							"color": {
								"text": "brown"
							}
						}
					}
				}
			}
		}
	}
}



```
</details>

7. Ensure the block style variation element styles are capped at `0-1-0` specificity
9. Test nesting block style variations and that the correct element styles are applied
10. Ensure block instance element style override the block style variation's element styles
11. Give other elements e.g. `heading` and `cite` a test, they should not have their selectors wrapped unless the style includes a pseudo selector
12. Give `button` and `caption` a run as well. Their selectors should be wrapped in `:root :where()` at all times.

## Screenshots or screencast 

https://github.com/WordPress/gutenberg/assets/60436221/73b8e1c2-a7e3-44a3-9847-fd1cd167fd86

### CODE REVIEW

**Comment by aaronrobertshaw** [comment]
I've flagged this as ready for review while I work on the backport, just in case there is another RC and we can sneak this fix in.

File: backport-changelog/6.6/7012.md [hunk 1]
```diff
@@ -0,0 +1,3 @@
+https://github.com/WordPress/wordpress-develop/pull/7012
+
+* https://github.com/WordPress/gutenberg/pull/63403
```
Comment (aaronrobertshaw) [inline backport-changelog/6.6/7012.md:3]: As this PR is intended for a 6.6 point release if there isn't an RC4, I've point the backport changelog under 6.6. If that needs changing let me know.

**Review by talldan — APPROVED** [review]
It's testing well for me following those steps, though I didn't dive deeply into testing variations.

**Review by andrewserong — APPROVED** [review]
This is testing great for me, thanks so much for the detailed testing instructions! The code change itself is straightforward and logical, but the extra detail in the testing instructions very much helps us be confident of the change here 👍

Here are my running testing notes:

✅ Custom HTML nav structure/markup is styled correct in post & site editors, and frontend
✅ Global element + hover styles correct in post & site editors, and frontend
✅ Global block type element styles correct in post & site editors, and frontend
✅ Element style overrides on block instances (e.g. individual paragraph block) correct in post & site editors, and frontend
✅ Block style variations including nested block element styles working correctly in post & site editors, and frontend
✅ Block style variation element styles are capped at `0-1-0`
✅ Nested block style variations reflect the nesting structure in post & site editors, and frontend
✅ Block instance element styles (e.g. a Group block's Button styles) override the block style variation's element styles
✅ Other elements, `heading` and `cite` work as expected (TIL the pullquote block has `color: inherit` that overrides the `cite` element styles — this is not an issue with this PR, just something I noticed)
✅ `button` and `caption` selectors are wrapped in `:root :where()` at all times.

LGTM! 🚀

**Comment by aaronrobertshaw** [comment]
There's a backport available for the PHP changes in https://github.com/WordPress/wordpress-develop/pull/7012

**Comment by andrewserong** [comment]
I just cherry-picked this PR to the Gutenberg release/18.8 branch to get it included in the next release: 32ff78fad3a

This will still need to be cherry picked for a core release.

**Comment by andrewserong** [comment]
A merge PR targeting the `wp/6.6` branch is open over in #63637.

**Comment by markhowellsmead** [comment]
Please urgently review the prioritisation issues being caused, in which inline styles are always overriding styles linked in external stylesheets. (Inline rules defined later in the cascade with the same specificity override externally-linked rules.)

- https://github.com/WordPress/gutenberg/issues/53468
- https://github.com/WordPress/gutenberg/issues/40159

**Comment by aaronrobertshaw** [comment]
Thanks for flagging this @markhowellsmead 👍 

There is an open ticket regarding the element styles issue for WP6.6. This [comment](https://core.trac.wordpress.org/ticket/61660#comment:16) outlines a custom plugin build that can be used to test this fix in the context of WP6.6. It would be great if you could try that and provide feedback on the ticket.

Another option if your site is using Gutenberg, is to update it to 18.8 which was released today and [contains this fix](https://github.com/WordPress/gutenberg/pull/63403#issuecomment-2232025769).

**Comment by markhowellsmead** [comment]
Thanks @aaronrobertshaw; I haven't used the Gutenberg plugin on our client sites yet because it's always broken quite a lot of things whenever I've tried to use it before, because of the rapid pace of development and because of its beta status. However, using 18.8.0 does fix the one CSS case I mentioned.

**Comment by aicuaboom** [comment]
📡🌏

**Comment by hellofromtonya** [comment]
>However, using 18.8.0 does fix the one CSS case I mentioned.

Thank you @markhowellsmead for testing and confirming  ⭐

**Comment by aicuaboom** [comment]
Autoadd all ipcdt648 nguyên thuỷ E3

**Comment by nickfmc** [comment]
Will this also address the padding over rides  on the body tag. if you have body{padding-top:150px} for say a fixed header the new root specificity will overide it

**Comment by talldan** [comment]
@nickfmc I'm working on a related fix in #63726, it may solve your issues. #63712 was the issue that was reported in relation to that.

**Pull request merged**